src/Microsoft.AspNetCore.Routing/Matching/Ascii.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    internal static class Ascii
+    {
+        // case-sensitive equality comparison when we KNOW that 'a' is in the ASCII range
+        // and we know that the spans are the same length.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool UnsafeAsciiIgnoreCaseEqual(ReadOnlySpan<char> a, ReadOnlySpan<char> b, int length)
+        {
+            // We're doing unsafe operations here, caller is expected to have already checked
+            // length.
+            ref var charA = ref MemoryMarshal.GetReference(a);
+            ref var charB = ref MemoryMarshal.GetReference(b);
+
+            while (length > 0 && AsciiIgnoreCaseEqual(charA, charB))
+            {
+                charA = ref Unsafe.Add(ref charA, 1);
+                charB = ref Unsafe.Add(ref charB, 1);
+                length--;
+            }
+
+            return length == 0;
+        }
+
+        // case-insensitive equality comparison for characters in the ASCII range
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool AsciiIgnoreCaseEqual(char charA, char charB)
+        {
+            const uint AsciiToLower = 0x20;
+            return
+                // Equal when chars are exactly equal
+                charA == charB ||
+
+                // Equal when converted to-lower AND they are letters
+                ((charA | AsciiToLower) == (charB | AsciiToLower) && (uint)((charA | AsciiToLower) - 'a') <= (uint)('z' - 'a'));
```

Comment:
What happens with non-alpha characeters? What will lower-casing a digit or `/`, `~` let it compare as equal to something else incorrectly?

Reply:
I just saw the comment that you check they are letters 😄

Reply:
How about removing the check for letters, writing some tests that fail with non-alpha characters, then add the check back. Make sure we have that situation covered test-wise 😸

Reply:
I think the test you're actually looking for here is a case like ` ` vs `@` these are 32 bits apart on the ASCII table so if something was wrong with line 42 we might accept them as equal when they are obviously not.

Reply:
Yes, exactly what I was thinking.